Migrate CanBeNonNullable to Analysis API

atulgpt opened the pull request:
Part of https://github.com/detekt/detekt/issues/8046

t-kameyama commented on detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:
```diff
@@ -504,67 +532,74 @@ class CanBeNonNullable(config: Config) :
-            } ?: false
+            val delegate = this ?: return true
+            return analyze(delegate) {
+                val functionSymbol = delegate
+                    .mainReference
+                    ?.resolveToSymbols()
+                    ?.filterIsInstance<KaFunctionSymbol>()
+                    ?.firstOrNull {
+                        it.callableId?.callableName == OperatorNameConventions.GET_VALUE
+                    }
+                functionSymbol?.run {
+                    if (returnType is KaTypeParameterType) {
+                        // todo<k2> ignoring some case which in pre k2 was passing as earlier
+                        //  using BindingContext.DELEGATED_PROPERTY_RESOLVED_CALL we were able to
+                        //  get the actual type of the implementation
+                        returnType.canBeNull
+                    } else {
+                        returnType.nullability != KaTypeNullability.NON_NULLABLE
+                    }
+                } ?: true
```
```suggestion
                val returnType = delegate
                    .mainReference
                    ?.resolveToSymbols()
                    ?.filterIsInstance<KaFunctionSymbol>()
                    ?.firstOrNull {
                        it.callableId?.callableName == OperatorNameConventions.GET_VALUE
                    }
                    ?.returnType
                    ?.let {
                        if (it is KaTypeParameterType) {
                            val classType = delegate.expression?.resolveToCall()?.singleFunctionCallOrNull()
                                ?.partiallyAppliedSymbol?.signature?.returnType as? KaClassType
                            classType?.typeArguments?.lastOrNull()?.type
                        } else {
                            it
                        }
                    }
                returnType?.nullability != KaTypeNullability.NON_NULLABLE
```

BraisGabin commented:
Does @t-kameyama's comment help with your issue?

If there is no way to implement it right now we can disable those TC and create an issue to track it (with a link to the issue on kotlin side that block us).

atulgpt commented:
> @atulgpt Please try the fix I proposed. All tests are passing.

Let me try the fix then

atulgpt commented:
> Does @t-kameyama's comment help with your issue?
> 
> If there is no way to implement it right now we can disable those TC and create an issue to track it (with a link to the issue on kotlin side that block us).

Hi @t-kameyama thanks for the comment. I tried your fix(I think something similar I explored and was not able to fix the cases). In your suggested comments following TCs are failing(It was not added earlier, I added it in my local)

```kt
import kotlin.reflect.KProperty

class A {
    private val a: Int? by PropDelegate<Int?, String>(0)
}

class PropDelegate<K: Any?, T: Any>(private var propVal: K) {
    operator fun getValue(thisRef: A, property: KProperty<*>): K {
        return propVal
    }
}
```

Above since we are mapping last type param to our analysis it is giving false positive(as that is not actually contributing to the type of the delegated value)

```kt
class A {
    private val a: Int? by PropDelegate<Int, String?>(0)
}

class PropDelegate<K: Any, T: Any?>(private var propVal: K) {
    operator fun getValue(thisRef: A, property: KProperty<*>): K {
        return propVal
    }
}
```

Same reason as above but constructed differently as to give false negative

```kt
class A(private var aDelegate: Int) {
    private val a: Int? by this::aDelegate
}
```

Here(a false negative) actual returnType is `KaTypeParameterType`(from main ref resolved call) but then it `delegate.expression` return type is `kotlin/Int` which doesn't have any `typeArguments`

@BraisGabin for now I have added the above TCs as well and marked those disabled(previously add TCs are passing with the @t-kameyama code though it is not identical to original pre analysis api code)

t-kameyama commented on detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullable.kt:
```diff
@@ -504,67 +535,80 @@ class CanBeNonNullable(config: Config) :
-                bindingContext[BindingContext.DELEGATED_PROPERTY_RESOLVED_CALL, it]
-                    ?.resultingDescriptor
+            val delegate = this ?: return true
+            return analyze(delegate) {
+                val returnType = delegate
```
Thank you very much for your review.
It seems that my proposed code could not handle delegate properties. Sorry for the extra trouble.
I think it would be better to revert to the original code to avoid any false positives.

BraisGabin commented on detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableSpec.kt:
```diff
@@ -86,9 +97,132 @@ class CanBeNonNullableSpec(val env: KotlinEnvironmentContainer) {
+            assertThat(subject.lintWithContext(env, code)).isEmpty()
+        }
+
+        @Disabled
```
Add a comment here

BraisGabin reviewed (commented):
Looks good to me. Only a missing message in one `@Disable`.

One question. Would those test work on the old `CanBeNonNullable` rule? If so it would be great to split this PR in two. One with only the test improvements (no `@Disable`). And then this one with the migration (and adding the `@Disable`).

atulgpt commented:
> Looks good to me. Only a missing message in one `@Disable`.
> 
> One question. Would those test work on the old `CanBeNonNullable` rule? If so it would be great to split this PR in two. One with only the test improvements (no `@Disable`). And then this one with the migration (and adding the `@Disable`).

Only one TC `does report when vals with getters uses unnecessary let` which passes here and fails in the old code. I will then add this TC in this PR

atulgpt commented:
> I keep getting this OOM trying to merge this PR. https://github.com/detekt/detekt/actions/runs/17238815254/job/48910101390?pr=8428
> 
> I added [8fb6fe3](https://github.com/detekt/detekt/commit/8fb6fe344620f2c310238ab94543e328d34f7ecc) to try to merge this.
> 
> cc @3flex because you are the one that is fine-tuning this value.

Thanks

3flex commented:
Unfortunately the memory issues are probably out of our hands so we'll have to keep bumping it.

Hoping we don't get a bunch of complaints when the alpha is out :/
